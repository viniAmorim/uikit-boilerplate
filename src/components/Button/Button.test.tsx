import React from 'react';

import snapshotDiff from 'snapshot-diff';

import { act, fireEvent, render } from '~/test/utils';

import { Button } from '..';

describe('Button', () => {
  it('should render correctly', () => {
    const { getByRole } = render(<Button>Button</Button>);

    const buttonElement = getByRole('button');

    expect(buttonElement).toBeInTheDocument();
    expect(buttonElement).toHaveTextContent(/^Button$/);
    expect(buttonElement).toHaveAttribute('type', 'button');
    expect(buttonElement).toBeEnabled();

    expect(buttonElement).toMatchSnapshot();
  });

  it('should render a primary button', () => {
    const { getByRole } = render(<Button variant="primary">Button</Button>);

    const buttonElement = getByRole('button');

    expect(buttonElement).toBeInTheDocument();

    expect(buttonElement).toMatchSnapshot();
  });

  it('should render a secondary button', () => {
    const { getByRole } = render(<Button variant="secondary">Button</Button>);

    const buttonElement = getByRole('button');

    expect(getByRole('button')).toBeInTheDocument();

    expect(buttonElement).toMatchSnapshot();
  });

  it('should render a button with border left variant', () => {
    const { getByRole } = render(<Button border="left">Button</Button>);

    const buttonElement = getByRole('button');

    expect(getByRole('button')).toBeInTheDocument();

    expect(buttonElement).toMatchSnapshot();
  });

  it('should render a button with border right variant', () => {
    const { getByRole } = render(<Button border="right">Button</Button>);

    const buttonElement = getByRole('button');

    expect(getByRole('button')).toBeInTheDocument();

    expect(buttonElement).toMatchSnapshot();
  });

  it('should render a button with different sizes', () => {
    const { getByRole } = render(<Button size="lg">Button</Button>);

    const buttonElement = getByRole('button');

    expect(getByRole('button')).toBeInTheDocument();

    expect(buttonElement).toMatchSnapshot();
  });

  it('should render a submit button', () => {
    const { getByRole } = render(<Button type="submit">Button</Button>);

    const buttonElement = getByRole('button');

    expect(buttonElement).toBeInTheDocument();
    expect(buttonElement).toHaveAttribute('type', 'submit');

    expect(buttonElement).toMatchSnapshot();
  });

  it('can be disabled', () => {
    const { rerender, asFragment, getByRole } = render(<Button>Button</Button>);

    const enabled = asFragment();

    rerender(<Button disabled>Button</Button>);

    const disabled = asFragment();

    expect(getByRole('button')).toBeDisabled();

    expect(disabled).toMatchSnapshot();
    expect(snapshotDiff(enabled, disabled)).toMatchSnapshot();
  });

  it('can be loading', () => {
    const { rerender, asFragment } = render(<Button>Button</Button>);

    const enabled = asFragment();

    rerender(<Button loading>Button</Button>);

    const loading = asFragment();

    expect(loading).toMatchSnapshot();
    expect(snapshotDiff(enabled, loading)).toMatchSnapshot();
  });

  it('can have onClick', () => {
    const handleClick = jest.fn();

    const { getByRole } = render(<Button onClick={handleClick}>Button</Button>);

    const buttonElement = getByRole('button');

    expect(buttonElement).toBeInTheDocument();

    fireEvent.click(buttonElement);

    expect(handleClick).toHaveBeenCalledTimes(1);
  });

  it('can have focus', () => {
    const { getByRole } = render(<Button>Button</Button>);

    const buttonElement = getByRole('button');

    expect(buttonElement).toBeInTheDocument();

    act(() => {
      fireEvent.keyDown(document.body, { key: 'TAB' });
      buttonElement.focus();
    });

    expect(buttonElement).toHaveFocus();

    expect(buttonElement).toMatchSnapshot();
  });
});
