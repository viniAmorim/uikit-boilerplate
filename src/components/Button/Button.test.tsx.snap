// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button can be disabled 1`] = `
<DocumentFragment>
  <button
    border="default"
    class="mantine-UnstyledButton-root mantine-Button-root sc-fsQiph gJlWZX mantine-1pjtuep"
    data-button="true"
    data-disabled="true"
    disabled=""
    type="button"
  >
    <div
      class="mantine-1wpc1xj mantine-Button-inner"
    >
      <span
        class="mantine-1ryt1ht mantine-Button-label"
      >
        Button
      </span>
    </div>
  </button>
</DocumentFragment>
`;

exports[`Button can be disabled 2`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -1,10 +1,12 @@
  <DocumentFragment>
    <button
      border=\\"default\\"
      class=\\"mantine-UnstyledButton-root mantine-Button-root sc-fsQiph gJlWZX mantine-1pjtuep\\"
      data-button=\\"true\\"
+     data-disabled=\\"true\\"
+     disabled=\\"\\"
      type=\\"button\\"
    >
      <div
        class=\\"mantine-1wpc1xj mantine-Button-inner\\"
      >"
`;

exports[`Button can be loading 1`] = `
<DocumentFragment>
  <button
    border="default"
    class="mantine-UnstyledButton-root mantine-Button-root sc-fsQiph gJlWZX mantine-1pjtuep"
    data-button="true"
    data-loading="true"
    type="button"
  >
    <div
      class="mantine-1wpc1xj mantine-Button-inner"
    >
      <span
        class="mantine-Button-icon mantine-Button-leftIcon mantine-1jbrfp"
      >
        <svg
          class="mantine-1avyp1d"
          height="calc(2.625rem / 2)"
          role="presentation"
          stroke="#fff"
          viewBox="0 0 38 38"
          width="calc(2.625rem / 2)"
          xmlns="http://www.w3.org/2000/svg"
        >
          <g
            fill="none"
            fill-rule="evenodd"
          >
            <g
              stroke-width="5"
              transform="translate(2.5 2.5)"
            >
              <circle
                cx="16"
                cy="16"
                r="16"
                stroke-opacity=".5"
              />
              <path
                d="M32 16c0-9.94-8.06-16-16-16"
              >
                <animatetransform
                  attributeName="transform"
                  dur="1s"
                  from="0 16 16"
                  repeatCount="indefinite"
                  to="360 16 16"
                  type="rotate"
                />
              </path>
            </g>
          </g>
        </svg>
      </span>
      <span
        class="mantine-1ryt1ht mantine-Button-label"
      >
        Button
      </span>
    </div>
  </button>
</DocumentFragment>
`;

exports[`Button can be loading 2`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -1,15 +1,58 @@
  <DocumentFragment>
    <button
      border=\\"default\\"
      class=\\"mantine-UnstyledButton-root mantine-Button-root sc-fsQiph gJlWZX mantine-1pjtuep\\"
      data-button=\\"true\\"
+     data-loading=\\"true\\"
      type=\\"button\\"
    >
      <div
        class=\\"mantine-1wpc1xj mantine-Button-inner\\"
+     >
+       <span
+         class=\\"mantine-Button-icon mantine-Button-leftIcon mantine-1jbrfp\\"
+       >
+         <svg
+           class=\\"mantine-1avyp1d\\"
+           height=\\"calc(2.625rem / 2)\\"
+           role=\\"presentation\\"
+           stroke=\\"#fff\\"
+           viewBox=\\"0 0 38 38\\"
+           width=\\"calc(2.625rem / 2)\\"
+           xmlns=\\"http://www.w3.org/2000/svg\\"
+         >
+           <g
+             fill=\\"none\\"
+             fill-rule=\\"evenodd\\"
+           >
+             <g
+               stroke-width=\\"5\\"
+               transform=\\"translate(2.5 2.5)\\"
              >
+               <circle
+                 cx=\\"16\\"
+                 cy=\\"16\\"
+                 r=\\"16\\"
+                 stroke-opacity=\\".5\\"
+               />
+               <path
+                 d=\\"M32 16c0-9.94-8.06-16-16-16\\"
+               >
+                 <animatetransform
+                   attributeName=\\"transform\\"
+                   dur=\\"1s\\"
+                   from=\\"0 16 16\\"
+                   repeatCount=\\"indefinite\\"
+                   to=\\"360 16 16\\"
+                   type=\\"rotate\\"
+                 />
+               </path>
+             </g>
+           </g>
+         </svg>
+       </span>
        <span
          class=\\"mantine-1ryt1ht mantine-Button-label\\"
        >
          Button
        </span>"
`;

exports[`Button can have focus 1`] = `
<button
  border="default"
  class="mantine-UnstyledButton-root mantine-Button-root sc-fsQiph gJlWZX mantine-1pjtuep"
  data-button="true"
  type="button"
>
  <div
    class="mantine-1wpc1xj mantine-Button-inner"
  >
    <span
      class="mantine-1ryt1ht mantine-Button-label"
    >
      Button
    </span>
  </div>
</button>
`;

exports[`Button should render a button with border left variant 1`] = `
<button
  border="left"
  class="mantine-UnstyledButton-root mantine-Button-root sc-fsQiph iRuqCZ mantine-1pjtuep"
  data-button="true"
  type="button"
>
  <div
    class="mantine-1wpc1xj mantine-Button-inner"
  >
    <span
      class="mantine-1ryt1ht mantine-Button-label"
    >
      Button
    </span>
  </div>
</button>
`;

exports[`Button should render a button with border right variant 1`] = `
<button
  border="right"
  class="mantine-UnstyledButton-root mantine-Button-root sc-fsQiph iebbtx mantine-1pjtuep"
  data-button="true"
  type="button"
>
  <div
    class="mantine-1wpc1xj mantine-Button-inner"
  >
    <span
      class="mantine-1ryt1ht mantine-Button-label"
    >
      Button
    </span>
  </div>
</button>
`;

exports[`Button should render a button with different sizes 1`] = `
<button
  border="default"
  class="mantine-UnstyledButton-root mantine-Button-root sc-fsQiph ekYqal mantine-1m1zbso"
  data-button="true"
  type="button"
>
  <div
    class="mantine-1wpc1xj mantine-Button-inner"
  >
    <span
      class="mantine-1ryt1ht mantine-Button-label"
    >
      Button
    </span>
  </div>
</button>
`;

exports[`Button should render a primary button 1`] = `
<button
  border="default"
  class="mantine-UnstyledButton-root mantine-Button-root sc-fsQiph gJlWZX mantine-1pjtuep"
  data-button="true"
  type="button"
>
  <div
    class="mantine-1wpc1xj mantine-Button-inner"
  >
    <span
      class="mantine-1ryt1ht mantine-Button-label"
    >
      Button
    </span>
  </div>
</button>
`;

exports[`Button should render a secondary button 1`] = `
<button
  border="default"
  class="mantine-UnstyledButton-root mantine-Button-root sc-fsQiph bSnkbv mantine-1pjtuep"
  data-button="true"
  type="button"
>
  <div
    class="mantine-1wpc1xj mantine-Button-inner"
  >
    <span
      class="mantine-1ryt1ht mantine-Button-label"
    >
      Button
    </span>
  </div>
</button>
`;

exports[`Button should render a submit button 1`] = `
<button
  border="default"
  class="mantine-UnstyledButton-root mantine-Button-root sc-fsQiph gJlWZX mantine-1pjtuep"
  data-button="true"
  type="submit"
>
  <div
    class="mantine-1wpc1xj mantine-Button-inner"
  >
    <span
      class="mantine-1ryt1ht mantine-Button-label"
    >
      Button
    </span>
  </div>
</button>
`;

exports[`Button should render correctly 1`] = `
<button
  border="default"
  class="mantine-UnstyledButton-root mantine-Button-root sc-fsQiph gJlWZX mantine-1pjtuep"
  data-button="true"
  type="button"
>
  <div
    class="mantine-1wpc1xj mantine-Button-inner"
  >
    <span
      class="mantine-1ryt1ht mantine-Button-label"
    >
      Button
    </span>
  </div>
</button>
`;
